flang/lib/Lower/OpenMP.cpp

```diff
@@ -135,6 +135,185 @@ static void createBodyOfOp(
+          }
+          break;
+        }
+      }
+    }
+    evalIter = (*evalIter).parentConstruct;
+  }
+  return orderedLevel;
+}
+
+static std::int64_t createOrderedDependSinkOp(
+    Fortran::lower::AbstractConverter &converter,
+    Fortran::lower::pft::Evaluation &eval,
+    const Fortran::parser::OmpDependClause::Sink *sinkVectors,
+    llvm::SmallVectorImpl<mlir::Value> &operands) {
+  auto &firOpBuilder = converter.getFirOpBuilder();
+  auto currentLocation = converter.getCurrentLocation();
+  Fortran::lower::StatementContext stmtCtx;
+  std::int64_t orderedLevel{0};
+
+  for (const auto &sinkVec : sinkVectors->v) {
+    mlir::Value varI32, sinkOperand;
+    auto ConstantOne = firOpBuilder.createIntegerConstant(
+        currentLocation, firOpBuilder.getI64Type(), 1);
+    // The name corresponds to the loop iterator and create its index.
+    const auto &name = std::get<Fortran::parser::Name>(sinkVec.t);
+    auto varSym = converter.getSymbolAddress(*name.symbol);
+    if (orderedLevel > 0) {
+      varI32 = firOpBuilder.create<fir::LoadOp>(currentLocation, varSym);
+    }
+    auto varI64 =
+        firOpBuilder.createConvert(currentLocation, firOpBuilder.getI64Type(),
+                                   orderedLevel > 0 ? varI32 : varSym);
+    auto varIndex = firOpBuilder.create<mlir::arith::SubIOp>(
+        currentLocation, varI64, ConstantOne);
+    // Analyze the "vec" in DEPEND(SINK: vec) if the offset exists such as
+    // depend(sink: j - 1), where "j" is the "name" and "1" is the scalar
+    // integer constant expression.
+    const auto &sinkVecLength =
+        std::get<std::optional<Fortran::parser::OmpDependSinkVecLength>>(
+            sinkVec.t);
+    if (sinkVecLength) {
+      const auto &definedOperator =
+          std::get<Fortran::parser::DefinedOperator>((*sinkVecLength).t);
+      const auto &intrinsicOperator =
+          std::get_if<Fortran::parser::DefinedOperator::IntrinsicOperator>(
+              &definedOperator.u);
+      const auto &expr =
+          std::get<Fortran::parser::ScalarIntConstantExpr>((*sinkVecLength).t);
+      auto offI32 = fir::getBase(
+          converter.genExprValue(*Fortran::semantics::GetExpr(expr), stmtCtx));
+      auto offI64 = firOpBuilder.createConvert(
+          currentLocation, firOpBuilder.getI64Type(), offI32);
+      if (*intrinsicOperator ==
+          Fortran::parser::DefinedOperator::IntrinsicOperator::Subtract) {
+        sinkOperand = firOpBuilder.create<mlir::arith::SubIOp>(
+            currentLocation, varIndex, offI64);
+      } else {
+        assert(*intrinsicOperator ==
+                   Fortran::parser::DefinedOperator::IntrinsicOperator::Add &&
+               "Operator must be Add or Subtract");
+        sinkOperand = firOpBuilder.create<mlir::arith::AddIOp>(
+            currentLocation, varIndex, offI64);
+      }
+    } else {
+      sinkOperand = varIndex;
+    }
+    operands.push_back(sinkOperand);
+    orderedLevel++;
```

Comment:
What is the relation between orderedLevels and sink vectors? Is there an assumption here?

Reply:
The orderedLevel is equal to the number of `sinkVec` in `sinkVectors`, which is the number of loops in the doacross loop. For example, `sinkVec` refers to `i` and `j` and `sinkVectors` refers to `(i, j)` for `!$omp ordered depend(sink: i, j)`. As OpenMP 5.1 2.19.11 states, `vec` has the form: x1 [+/- d1], x2 [+/- d2], . . . , xn [+/- dn], where n is the value specified by the ordered clause in the worksharing-loop directive. When the depend type is `sink`, the loop iterators can be obtained from the `vec`, and it is not necessary to find this info from the ordered clause in worksharing-loop directive.

Reply:
I had one patch under review to include a semantic check for this. The error message is emitted as "The number of variables in DEPEND(SINK: vec) clause does not match the parameter specified in ORDERED clause" in https://reviews.llvm.org/D113399.
For `depend(sink: vec)` clause, finding the wsloop directive is not necessary, so its info is unknown.